File: 060/001/038.py
#!/usr/bin/env python3
"""
关卡 3-1 · LangChain 重写（填空版）

核心：把 1-10 的综合单 Agent（full_agent.py，约 160 行手写：工具 schema dict + 函数分发 dict
+ 重试/超时/幂等/降级 + tracing + 手写 while 循环）用 LangChain 重写，看代码量缩到多少、
少了哪些样板。

执行流程图（python3 038.py）：

用户问题
  └─ agent_executor.invoke({"input": 问题})
       └─ LangChain 内置 ReAct 循环（不用你手写 while）：
            ├─ 带工具的 LLM 决定要不要调工具
            ├─ 要 → 执行 @tool 装饰的函数 → 结果自动喂回
            └─ 不要 → 返回最终答案

LangChain 帮你省掉的样板：
  1. TOOLS 的 JSON schema 声明  → @tool 装饰器自动从函数签名 + docstring 生成
  2. TOOL_FUNCS 字典 + 按名分发 → 框架自动绑定
  3. 手写 while 循环 / 消息组装 → AgentExecutor 内置
LangChain 不替你写的（还得自己补）：
  重试/超时/幂等/降级、tracing、防护 —— 这些「生产细节」框架不管。

依赖（先装，装不上先看代码理解概念）：
  pip3 install langchain langchain-openai

规则：
  1. 下面有 3 个空（填空 1 含 a/b 两处），填对了才能跑通。
  2. 卡住看 answers-038-042.md（关卡 3-1 那节）。
  3. 跑通：打印出「🎯 最终答案:」那一行。
"""

from langchain_openai import ChatOpenAI
from langchain_core.tools import tool
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain.agents import create_tool_calling_agent, AgentExecutor


# ── 填空 1 ──────────────────────────────────────
# 创建连 DeepSeek 的 ChatOpenAI。DeepSeek 是 OpenAI 兼容接口，所以用 ChatOpenAI 也能连。
# 提示：base_url 填 DeepSeek 的 OpenAI 兼容地址，model 填 deepseek-chat
llm = ChatOpenAI(
    model="________",                    # ← 填空 1a：模型名
    api_key="sk-你的key",                # ← 换成你自己的 key
    base_url="________",                 # ← 填空 1b：DeepSeek 的 OpenAI 兼容地址
    temperature=0,
)


# 工具：用 @tool 装饰器，schema（name/description/参数）自动从函数签名和 docstring 生成
@tool
def get_weather(city: str) -> str:
    """查询某个城市的天气。"""
    return f"{city} 今天晴，25 度（模拟）"


@tool
def calc(expr: str) -> str:
    """计算数学表达式，如 3*(5+2)。"""
    allowed = set("0123456789+-*/(). ")
    if any(c not in allowed for c in expr):
        return "表达式含非法字符"
    return str(eval(expr))


# ── 填空 2 ──────────────────────────────────────
# 拼 prompt：system 给角色，两个 MessagesPlaceholder 放「对话历史」和「中间工具往返记录」。
# 提示：tool-calling agent 必须有一个 agent_scratchpad 占位，框架往里塞「工具调用→结果」的往返
prompt = ChatPromptTemplate.from_messages([
    ("system", "你是一个助手。查天气用 get_weather，算数用 calc。回答用中文。"),
    MessagesPlaceholder("chat_history"),
    ("human", "{input}"),
    MessagesPlaceholder("________"),      # ← 填空 2：中间工具往返记录的占位名
])


# ── 填空 3 ──────────────────────────────────────
# 组装 agent：把「带工具的 LLM + prompt」变成工具调用 agent，再包一层执行器
# 提示：create_tool_calling_agent 三个参数是 (llm, tools, prompt)，tools 是 [get_weather, calc]
agent = create_tool_calling_agent(
    ________,        # ← 填空 3a：LLM
    ________,        # ← 填空 3b：工具列表
    ________,        # ← 填空 3c：prompt
)
agent_executor = AgentExecutor(agent=agent, tools=[get_weather, calc], verbose=True)


if __name__ == "__main__":
    print("=" * 55)
    print("  LangChain 重写 1-10 综合 Agent（对比代码量）")
    print("=" * 55)
    # 和 1-10 一样的问题，看 LangChain 版多简洁
    result = agent_executor.invoke({
        "input": "北京天气怎么样？顺便算一下 3*(5+2) 等于多少",
        "chat_history": [],
    })
    print("\n🎯 最终答案:", result["output"])

File: 060/001/038.go
package main

/*
关卡 3-1 · Go 版 · 工具调用 Agent（对比 Python 版 038.py 的 LangChain 重写）

核心：Python 里用 LangChain 把 1-10 手写的 160 行综合 agent 缩到 ~40 行——
@tool 装饰器从函数签名自动生成 schema、AgentExecutor 内置 ReAct 循环。
Go 没有 LangChain 生态，这里手写一个「极简工具调用 agent」，把 Python 里
被框架藏起来的样板（工具表 + 按名分发 + ReAct 循环）显式写出来——
正好让你看清 LangChain 到底替你省了什么。

执行流程（go run 038.go）：
  用户问题 → LLM 决定要调工具 → get_weather("北京") → calc("3*(5+2)")
           → 结果喂回 → LLM 输出最终答案

Go 和 Python 最大的不同：
  Python 的 @tool 装饰器自动生成 schema；Go 没有装饰器，工具 = struct（name + description + run 函数）。
  Python 的 eval("3*(5+2)") Go 里没有，得手写递归下降表达式求值器（evalExpr）。
  Python 的 ReAct while 循环由 AgentExecutor 内置；Go 手写（agentLoop）。
  本演示离线可跑：不真调 LLM，用「硬编码的调用计划」模拟 LLM 的决策。
*/

import (
	"fmt"
	"strconv"
	"strings"
)

// Tool：一个工具 = 名字 + 描述 + 执行函数（Go 版的 @tool）
type Tool struct {
	name        string
	description string
	run         func(args map[string]string) string
}

var tools = []Tool{
	{"get_weather", "查询某个城市的天气", func(a map[string]string) string {
		return fmt.Sprintf("%s 今天晴，25 度（模拟）", a["city"])
	}},
	{"calc", "计算数学表达式，如 3*(5+2)", func(a map[string]string) string {
		expr := a["expr"]
		if !allowedExpr(expr) {
			return "表达式含非法字符"
		}
		v, err := evalExpr(expr)
		if err != nil {
			return "表达式错误"
		}
		return strconv.FormatFloat(v, 'f', -1, 64)
	}},
}

// findTool：按名字找工具（对应 Python 里框架自动绑定的「按名分发」）
func findTool(name string) (Tool, bool) {
	for _, t := range tools {
		if t.name == name {
			return t, true
		}
	}
	return Tool{}, false
}

// agentLoop：极简 ReAct 循环 —— 模拟 LLM 的决策：先调工具，再给最终答案
// （Python 里这一步是 AgentExecutor 内置的，Go 里手写）
func agentLoop(question string) string {
	// 模拟 LLM 的决策：识别出两个工具调用（真系统里这是 LLM 输出的 function call）
	type call struct {
		tool string
		args map[string]string
	}
	calls := []call{
		{"get_weather", map[string]string{"city": "北京"}},
		{"calc", map[string]string{"expr": "3*(5+2)"}},
	}

	var results []string
	for _, c := range calls {
		t, ok := findTool(c.tool)
		if !ok {
			results = append(results, fmt.Sprintf("未知工具 %s", c.tool))
			continue
		}
		out := t.run(c.args)
		results = append(results, out)
		fmt.Printf("  🔧 调用 %s(%v) → %s\n", c.tool, c.args, out)
	}

	// 模拟 LLM 拿到工具结果后拼出最终答案
	return fmt.Sprintf("%s；3*(5+2) = %s", results[0], results[1])
}

func main() {
	fmt.Println("=======================================================")
	fmt.Println("  工具调用 Agent（对比 038.py 的 LangChain 重写）")
	fmt.Println("=======================================================")
	q := "北京天气怎么样？顺便算一下 3*(5+2) 等于多少"
	fmt.Printf("\n用户问题：%s\n\n", q)

	answer := agentLoop(q)
	fmt.Printf("\n🎯 最终答案: %s\n", answer)
	fmt.Println("\n（LangChain 在 Python 里替你省掉了：工具表 + 按名分发 + ReAct 循环，")
	fmt.Println("  Go 里这些都手写出来，正是 038.py 里框架藏起来的样板）")
}

// ============ 表达式求值器（对应 Python 的 eval）============

// allowedExpr：只允许数字和 + - * / ( ) . 空格
func allowedExpr(s string) bool {
	for _, c := range s {
		if !strings.ContainsRune("0123456789+-*/(). ", c) {
			return false
		}
	}
	return true
}

type exprParser struct {
	s   string
	pos int
}

// parseExpr：加减（最低优先级）
func (p *exprParser) parseExpr() (float64, error) {
	v, err := p.parseTerm()
	if err != nil {
		return 0, err
	}
	for p.pos < len(p.s) {
		op := p.s[p.pos]
		if op != '+' && op != '-' {
			break
		}
		p.pos++
		rhs, err := p.parseTerm()
		if err != nil {
			return 0, err
		}
		if op == '+' {
			v += rhs
		} else {
			v -= rhs
		}
	}
	return v, nil
}

// parseTerm：乘除（中优先级）
func (p *exprParser) parseTerm() (float64, error) {
	v, err := p.parseFactor()
	if err != nil {
		return 0, err
	}
	for p.pos < len(p.s) {
		op := p.s[p.pos]
		if op != '*' && op != '/' {
			break
		}
		p.pos++
		rhs, err := p.parseFactor()
		if err != nil {
			return 0, err
		}
		if op == '*' {
			v *= rhs
		} else {
			v /= rhs
		}
	}
	return v, nil
}

// parseFactor：数字或括号表达式（最高优先级）
func (p *exprParser) parseFactor() (float64, error) {
	if p.pos >= len(p.s) {
		return 0, fmt.Errorf("表达式不完整")
	}
	if p.s[p.pos] == '(' {
		p.pos++
		v, err := p.parseExpr()
		if err != nil {
			return 0, err
		}
		if p.pos >= len(p.s) || p.s[p.pos] != ')' {
			return 0, fmt.Errorf("缺少右括号")
		}
		p.pos++
		return v, nil
	}
	start := p.pos
	for p.pos < len(p.s) && p.s[p.pos] >= '0' && p.s[p.pos] <= '9' {
		p.pos++
	}
	if start == p.pos {
		return 0, fmt.Errorf("非法数字")
	}
	return strconv.ParseFloat(p.s[start:p.pos], 64)
}

func evalExpr(s string) (float64, error) {
	s = strings.ReplaceAll(s, " ", "")
	p := &exprParser{s: s}
	v, err := p.parseExpr()
	if err != nil {
		return 0, err
	}
	if p.pos != len(s) {
		return 0, fmt.Errorf("表达式尾部有多余字符")
	}
	return v, nil
}
